src/bootupd.rs

```diff
@@ -489,6 +490,149 @@ pub(crate) fn client_run_validate() -> Result<()> {
     Ok(())
 }
 
+pub(crate) fn client_run_migrate() -> Result<()> {
+    // Used to condition execution of this unit at the systemd level
+    let stamp_file = "/boot/.bootupd-static-migration-complete";
+
+    // Did we already complete the migration?
+    let mut ostree_cmd = std::process::Command::new("ostree");
+    let result = ostree_cmd
+        .args([
+            "config",
+            "--repo=/sysroot/ostree/repo",
+            "get",
+            "sysroot.bootloader",
+        ])
+        .output()
+        .with_context(|| "failed to call ostree command. Not performing migration")?;
+    if !result.status.success() {
+        // ostree will exit with a non zero return code if the key does not exists
+        println!("ostree repo 'sysroot.bootloader' config option not set yet.");
+    } else {
+        let bootloader = String::from_utf8(result.stdout)
+            .with_context(|| "decoding as UTF-8 output of ostree command")?;
+        if bootloader.trim_end() == "none" {
+            println!("ostree repo 'sysroot.bootloader' config option already set to 'none'.");
+            println!("Assuming that the migration is already complete.");
+            File::create(stamp_file)?;
+            return Ok(());
+        } else {
+            println!(
+                "ostree repo 'sysroot.bootloader' config currently set to: {}",
+                bootloader.trim_end()
+            );
+        }
+    }
+
+    // Remount /boot read write just for this unit (we are called in a slave mount namespace by systemd)
+    ensure_writable_boot()?;
+
+    let grub_config_dir = PathBuf::from("/boot/grub2");
+    let Ok(dirfd) = openat::Dir::open(&grub_config_dir).with_context(|| "a") else {
+        anyhow::bail!(
+            "Could not open {}. Not performing migration.",
+            grub_config_dir.display()
+        );
+    };
```

Comment:
Here and elsewhere I think the translation from "careful error checking shell" to Rust doesn't make sense because we can just use Rust's builtin `?`. Like this

```
let dirfd = Dir::open(&grub_config_dir).context("Opening /boot/grub2")?;
```

The "Not performing migration" bit is implicit in the whole operation.